builtins/web/blob.cpp

```diff
@@ -0,0 +1,499 @@
+  if (!promise) {
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  args.rval().setObject(*promise);
+
+  auto buffer = data_to_owned_array_buffer(cx, self);
+  if (!buffer) {
+    return RejectPromiseWithPendingError(cx, promise);
+  }
+
+  JS::RootedValue result(cx);
+  result.setObject(*buffer);
+  JS::ResolvePromise(cx, promise, result);
+
+  return true;
+}
+
+bool Blob::bytes(JSContext *cx, unsigned argc, JS::Value *vp) {
+  METHOD_HEADER(0)
+
+  JS::RootedObject promise(cx, JS::NewPromiseObject(cx, nullptr));
+  if (!promise) {
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  args.rval().setObject(*promise);
+
+  JS::RootedObject buffer(cx, data_to_owned_array_buffer(cx, self));
+  if (!buffer) {
+    return RejectPromiseWithPendingError(cx, promise);
+  }
+
+  auto len = JS::GetArrayBufferByteLength(buffer);
+  JS::RootedObject byte_array(cx, JS_NewUint8ArrayWithBuffer(cx, buffer, 0, len));
+  if (!byte_array) {
+    return RejectPromiseWithPendingError(cx, promise);
+  }
+
+  JS::RootedValue result(cx);
+  result.setObject(*byte_array);
+  JS::ResolvePromise(cx, promise, result);
+
+  return true;
+}
+
+bool Blob::slice(JSContext *cx, unsigned argc, JS::Value *vp) {
+  METHOD_HEADER(0)
+
+  auto src = Blob::blob(self);
+  int64_t size = src->size();
+  int64_t start = 0;
+  int64_t end = size;
+
+  JS::RootedString contentType(cx, JS_GetEmptyString(cx));
+
+  if (args.hasDefined(0)) {
+    HandleValue start_val = args.get(0);
+    if (!JS::ToInt64(cx, start_val, &start)) {
+      return false;
+    }
+  }
+
+  if (args.hasDefined(1)) {
+    HandleValue end_val = args.get(1);
+    if (!JS::ToInt64(cx, end_val, &end)) {
+      return false;
+    }
+  }
+
+  if (args.hasDefined(2)) {
+    HandleValue contentType_val = args.get(2);
+    if (!(contentType = normalize_type(cx, contentType_val))) {
+      return false;
+    }
+  }
+
+  // A negative value, is treated as an offset from the end of the Blob toward the beginning
+  start = (start < 0) ? std::max((size + start), 0LL) : std::min(start, size);
+  end = (end < 0) ? std::max((size + end), 0LL) : std::min(end, size);
+
+  auto dst = (end - start > 0)
+    ? std::make_unique<std::vector<uint8_t>>(src->begin() + start, src->begin() + end)
+    : std::make_unique<std::vector<uint8_t>>();
+
+  JS::RootedObject new_blob(cx, create(cx, std::move(dst), contentType));
+  if (!new_blob) {
+    return false;
+  }
+
+  args.rval().setObject(*new_blob);
+  return true;
+}
+
+bool Blob::text(JSContext *cx, unsigned argc, JS::Value *vp) {
+  METHOD_HEADER(0)
+
+  JS::RootedObject promise(cx, JS::NewPromiseObject(cx, nullptr));
+  if (!promise) {
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  args.rval().setObject(*promise);
+
+  auto blob = Blob::blob(self);
+  auto chars = JS::UTF8Chars((char *)blob->data(), blob->size());
+
+  JS::RootedString str(cx, JS_NewStringCopyUTF8N(cx, chars));
```

Comment:
The return value should be a result of decoding UTF8 on the blob as per https://w3c.github.io/FileAPI/#dom-blob-text. It looks like decoding function is fallible which makes some of the tests to fail. 

For example this test: https://github.com/web-platform-tests/wpt/blob/master/FileAPI/blob/Blob-text.any.js#L45